README.md
# oVirt Desktop Client

### Description

This project is a simple oVirt-related desktop client. The application will communicate with oVirt via its Python API and provide a list of VMs and VmPools on which the user has usage privileges, allowing two essential operations:

 1. _Manage power_: Users will be able to shutdown and start their machines.
 2. _Connect to the machine_: The application makes use of _virt-viewer_ to connect to the VM. Both VNC and SPICE protocols are supported.

This application is written in *PyQT5*, and has been verified to be working with oVirt version >= 4.0.0. It won't work with 3.6.0 versions of oVirt correctly.

This project is also accessible from https://ovirt-desktop-client.click

**Disclaimer**: This is an **unofficial** oVirt-related project. 

### Installation

As of version 2.0.0 of oVirt Desktop Client, the installation process is much simpler, because Python3 allows using native PyQt5-sip and PyQt libraries. In previous versions this had to be built manually.

Below you'll find the installation steps. You can also find a [Troubleshooting guide](https://github.com/nkovacne/ovirt-desktop-client/wiki/Install-process). If you find a different problem please feel free to report it.

On Debian/Ubuntu environments, make sure to install these packages: 
```
python3 python3-dev python3-virtualenv python-virtualenv qt5-default libcurl4-openssl-dev libxml2 libxml2-dev libxslt1-dev libssl-dev virt-viewer
```

On RHEL environments, make sure to install these packages:
```
python36 python36-devel python36-virtualenv python36-pip gcc gcc-c++ qt5-qtbase qt5-qtbase-devel libcurl-devel libxml2 libxml2-devel libxslt-devel openssl-devel virt-viewer
```

**Important for RHEL based environments**: Please note that in further steps, commands may vary in RHEL based environments. For example, it's likely your virtualenv command will be `virtualenv-3.6`, and your pip command will be `pip-3.6`. Please find the correct command in your case.

Now just clone, install a virtualenv with Python3 as the Python executable and install the requirements for the project.

```
git clone https://github.com/nkovacne/ovirt-desktop-client.git
virtualenv -p python3 --always-copy venv
. venv/bin/activate
cd ovirt-desktop-client
pip install -r requirements.txt
```

### Upgrading from version 1.x to 2.x

In previous versions, SIP and PyQt packages needed to be built manually. Furthermore, it was based on Python 2, whereas version >= 2.0.0 is Python 3 based. To upgrade, the fastest way is to `pull` the 2.0.0 version, destroy the virtualenv and recreate it. Just follow these commands to upgrade.

```
deactivate (in case you're working in the virtualenv)
rm -rf venv
virtualenv -p python3 --always-copy venv
. venv/bin/activate
cd ovirt-desktop-client
git pull
pip install -r requirements.txt
```

Check the `settings.conf.example` file just to check if you have all parameters up to date, and add those missing in your `settings.conf` file.

### Configuration

To run the application, a file named `settings.conf` must exist in the same directory where the application resides. This file contains 2 sections with a few parameters in each. You can find a `settings.conf.example` file inside the repository so you can base your configuration on it (don't forget to copy/rename it to `settings.conf`).

#### ovirt section

The beggining of this section is marked with the `[ovirt]` line and references some settings that are directly related to the oVirt infrastructure that you mean to connect to. It only has 2 parameters, and both are **mandatory**:

 * **url**: You oVirt infrastructure API URL. If you're using oVirt version 4.0.x or greater, URL should be somewhat like: `https://myovirt.mydomain.com/ovirt-engine/api`.
 * **domain**: The domain under which your users will authenticate. When you create an AAA authenticator (LDAP, Kerberos, ...), a domain name is created to match it. This value goes here, so users will authenticate as `username@domain` (Ex: LDAP, MyCompany, ...). It's the 'Profile' field value when you're logging into the oVirt web-based API.
 
#### app section

The beggining of this section is marked with the `[app]` line and references some settings that are directly related to the oVirt desktop client behavior. All of them are **optional** but have some default values.

* **lang**: Chooses the application language. Available languages are stored under the `lang` folder. If you don't see your language, you can translate it and send a push request so I can merge it into the project. Default: en.
* **connection_timeout**: Establishes the number of seconds to wait after sending a request to your oVirt infrastructure after which the request will be considered timed out. Default: 15
* **preferred_protocol**: Some VMs have more than one graphics protocol enabled (i.e, VNC and SPICE) at a time. These VMs allow choosing which one you prefer to use when opening a console window. In case that the VM only has one protocol enabled, this will be used disregarding this parameter. Possible values: spice, vnc. Default: spice
* **fullscreen**: Enables opening the console window as fullscreen or the default size. Possible values: 1 (full screen), 0 (default size). Default: 0
* **allow_remember**: Defines whether show or not the "Remember credentials" checkbox. Some environments might prefer not enabling it (for example, thin-client fashioned terminals). Possible values: 1 (enable the remembering credentials checkbox), 0 (disable the remembering credentials checkbox). Default: 1
* **autologout**: Time in minutes of idleness after which the session is forcibly closed. If set to 0, the autologout feature is disabled. Only works if there are not any credentials stored (also with allow_remember = 0), which is incompatible with this feature. Default: 0
* **notify_autologout**: The time before *autologout* in minutes before a warning window will be shown to the user alerting them about an imminent forced logout event. Accepting the warning means resetting the idle time. This setting needs to have a lower value than *autologout*. If this setting is set and 'autologout' is not, or if the value of *notify_autologout* is lower than the value in *autologout*, this setting will be set to the default value. A value of 0 means that no warning windows will be shown to the user. _Example_: If *autologout* is 15 and *notify_autologout* is 5, means that 5 minutes before reaching the 15 minutes limit of idleness a warning window will be shown. If the user accepts the warning within the next 5 minutes limit, the idleness count will be reset. Otherwise the enforced logout will be performed. Default: 0
* **remote_viewer_path**: The path to the remote-viewer binary. By default, it's set to a path that is compatible with most systems. However, you can set a customized path here. If set to an invalid path, the app will still try to find the correct binary. Will exit if no suitable binary was found. Default: /usr/bin/remote-viewer

### How to run

Just run the `virtualenv` and the `ovirtclient.py` executable.

```
. venv/bin/activate
cd ovirt-desktop-client
python ovirtclient.py
```

### Current version

Current stable version is 2.0.0. You can find a CHANGELOG file inside your directory to see news.


CHANGELOG
2.0.0
-----
**MAJOR**: For migration process for 1.x to 2.0.0 please see README.md.

* Migration to Python 3.x
* Migration to oVirt Python SDK 4.0

1.0.6b
------
* Hotfix of a DBZ when no VMs or VMPools are assigned to a user

1.0.6
-----
* Implemented password obfuscation. It makes no sense trying to encrypt credentials in a Python code as anything can be decrypted, decompiled, etc., so at least password is now obfuscated if stored.

  **IMPORTANT**: If you already stored your credentials before this version, they won't match once you run your client after pulling latest changes, so expect an authentication error. Either enter the credentials again and select the "store my credentials" checkbox or remove the credentials file yourself (~/.ovirtclient).

1.0.5-2
-------
* Added the 'reboot in progress' event
* Fixed some uncaught exceptions

1.0.5-1
-------
* Fix: _create_unverified_context() seems not to be present in some older versions of Python, replaced it with the proper way (creating an empty context for urlopen)
* Improved installation instructions in README

1.0.5
-----
* Implemented multithreading for initiating more than one viewer (improves UX as well)
* Added the app->remote_viewer_path setting
* Minor fixes

1.0.4
-----
* Stripped the HTML tags on authentication errors
* An autologout policy has been implemented, read README.md for more info about *autologout* and *notify_autologout* parameters.
* Parameters documented in settings.conf.example as well
* Some minor fixes

1.0.3
-----
* Added the VmPool listing support
* Fixed some uncaught exceptions

1.0.2
-----
* Forced the ovirt-engine-sdk-python version to 3.6.9 (4.0.X oVirt format is completely incompatible with 3.X.X versions)
* Fixed some issues with self-signed certificates (also affects LetsEncrypt certificates for some reason)
* Adeed the app->allow_remember configuration parameter, defining whether the "Remember credentials" checkbox is or not shown. Defaults 1 (allow remembering).
* Added a TODOLIST

1.0.1
-----
* Some minor fixes

1.0.0
-----
* Initial version, feel free to report any bugs you may encounter.


about.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# This file may be used under the terms of the GNU General Public License
# version 3.0 as published by the Free Software Foundation and appearing in
# the file LICENSE included in the packaging of this file.  Please review the
# following information to ensure the GNU General Public License version 3.0
# requirements will be met: http://www.gnu.org/copyleft/gpl.html.
#
# If you do not wish to use this file under the terms of the GPL version 3.0
# then you may purchase a commercial license.  For more information contact
# info@riverbankcomputing.com.
#
# This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
# WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.

import gettext
from version import VERSION
from PyQt5.QtWidgets import QPushButton, QDesktopWidget, QDialog, QLabel, QGridLayout
from PyQt5.QtGui import QImage, QPixmap 
from PyQt5.QtCore import Qt

class About(QDialog):
    """
        This class shows additional information about the application
    """

    def __init__(self, parent=None):
        QDialog.__init__(self, parent)
        self.initUI()

    def initUI(self):
        """
            Description: Simply shows the dialog
            Arguments: None
            Returns: Nothing
        """

        global conf

        self.resize(450, 150)

        # About image
        filename = 'imgs/about.png'
        image = QImage(filename)
        imageLabel = QLabel()
        imageLabel.setPixmap(QPixmap.fromImage(image))
        imageLabel.setAlignment(Qt.AlignCenter)

        # Labels for info
        lab_appname = QLabel("<font color='#0000FF'>" + _('apptitle') + ' ' + VERSION + "</font>")
        lab_appname.setAlignment(Qt.AlignLeft | Qt.AlignVCenter)
        lab_desc = QLabel(_('appdesc'))
        lab_desc.setAlignment(Qt.AlignLeft | Qt.AlignVCenter)
        lab_desc.setWordWrap(True)
        lab_author = QLabel(_('written_by') + ' nKn (<a href=\'http://github.com/nkovacne\'>http://github.com/nkovacne</a>)')
        lab_author.setAlignment(Qt.AlignLeft | Qt.AlignVCenter)
        lab_unoff = QLabel('<b>' + _('unofficial_project') + '</b>')
        lab_unoff.setAlignment(Qt.AlignLeft | Qt.AlignVCenter)
        lab_unoff.setWordWrap(True)

        # OK button
        okButton = QPushButton(_("ok"))
        okButton.setMaximumWidth(100)
        okButton.setDefault(True)
        okButton.clicked.connect(self.done)

        # Grid layout with all the elements
        grid = QGridLayout()

        grid.addWidget(imageLabel, 1, 0, 4, 1)               # About image

        grid.addWidget(lab_appname, 1, 1, 1, 2)              # Several QLabels
        grid.addWidget(lab_desc, 2, 1, 1, 2)
        grid.addWidget(lab_author, 3, 1, 1, 2)
        grid.addWidget(lab_unoff, 4, 1, 1, 2)

        grid.addWidget(okButton, 6, 1)                       # Button
        
        self.setLayout(grid) 

        self.setModal(True)
        self.center()
        self.setWindowTitle(_('about'))
        self.show()

    def center(self):
        """
            Description: Just centers the window
            Arguments: None
            Returns: Nothing
        """

        qr = self.frameGeometry()
        cp = QDesktopWidget().availableGeometry().center()
        qr.moveCenter(cp)
        self.move(qr.topLeft())


globalconf.py
#!/usr/bin/env python

# This file may be used under the terms of the GNU General Public License
# version 3.0 as published by the Free Software Foundation and appearing in
# the file LICENSE included in the packaging of this file.  Please review the
# following information to ensure the GNU General Public License version 3.0
# requirements will be met: http://www.gnu.org/copyleft/gpl.html.
#
# If you do not wish to use this file under the terms of the GPL version 3.0
# then you may purchase a commercial license.  For more information contact
# info@riverbankcomputing.com.
#
# This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
# WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.

from os.path import expanduser

class Configs:
    CONFIGFILE='settings.conf'
    USERCREDSFILE=expanduser('~') + '/.ovirtclient'
    USERNAME=None
    PASSWORD=None
    OVIRTCONN=None
    CONFIG={}
conf = Configs()

IMGDIR = 'imgs/'
UPDATESLEEPINTERVAL = 5
MAXWIDTH = 500
MAXHEIGHT = 600
BACKGROUNDCSS = 'background: black; color: white'
STANDARDCELLCSS = 'background: #1a1a1a; padding: 5px;'


credentials.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# This file may be used under the terms of the GNU General Public License
# version 3.0 as published by the Free Software Foundation and appearing in
# the file LICENSE included in the packaging of this file.  Please review the
# following information to ensure the GNU General Public License version 3.0
# requirements will be met: http://www.gnu.org/copyleft/gpl.html.
#
# If you do not wish to use this file under the terms of the GPL version 3.0
# then you may purchase a commercial license.  For more information contact
# info@riverbankcomputing.com.
#
# This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
# WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.

import gettext
import os
import configparser
from re import sub
from codecs import encode, decode
from os.path import isfile
from globalconf import conf, IMGDIR
from PyQt5.QtWidgets import QProgressBar, QPushButton, QDesktopWidget, QDialog, QLabel, QLineEdit, QGridLayout, QCheckBox, QMessageBox
from PyQt5.QtGui import QImage, QPixmap, QIcon
from PyQt5.QtCore import QBasicTimer, Qt
from ovirtsdk4 import Connection, Error

class CheckCreds(QDialog):
    """
        This class manages the authentication process for oVirt. If credentials are saved, they're
        tried automatically. If they are wrong, the username/password dialog is shown.
    """

    def __init__(self, parent, username, password, remember):
        QDialog.__init__(self, parent)
        self.uname = username
        self.pw = password
        self.remember = False if conf.CONFIG['allow_remember'] == '0' else remember
        self.setModal(True)

        self.initUI()

    def initUI(self):
        """
            Description: A progress bar, a status message will be shown and a timer() method will
                         be invoked as part of the authentication process.
            Arguments: None
            Returns: Nothing
        """

        self.pbar = QProgressBar(self)
        self.pbar.setGeometry(30, 40, 200, 25)
        
        self.status = QLabel(self)
        self.status.setGeometry(30, 75, 200, 20)

        self.timer = QBasicTimer()
        self.step = 0
        
        self.setGeometry(300, 300, 255, 100)
        self.center()
        self.setWindowTitle(_('loading'))
        self.show()

        self.timer.start(100, self)
    
    def timerEvent(self, e):
        """
            Description: Called periodically as part of the QBasicTimer() object.
                         Authentication will be handled within this method.
            Arguments: The event. Won't be used, though.
            Returns: Nothing, just exits when progress bar reaches 100%
        """

        global conf

        err = QMessageBox()
        self.status.setText(_('authenticating'))

        if not conf.USERNAME:
            try:
                conn = Connection(
                  url=conf.CONFIG['ovirturl'],
                  username=self.uname + '@' + conf.CONFIG['ovirtdomain'],
                  password=self.pw,
                  ca_file=conf.CONFIG['cafile'],
                  insecure=True,
                  timeout=int(conf.CONFIG['conntimeout']),
                  headers={'filter':True}
                )

                conn.test(raise_exception=True)

                conf.SOCKOBJ = conn
                conf.OVIRTCONN = conn.system_service()
                conf.USERNAME = self.uname
                conf.PASSWORD = self.pw
                self.status.setText(_('authenticated_and_storing'))
                self.step = 49
            except Error as e:
                err.critical(self, _('apptitle') + ': ' + _('error'), _('ovirt_connection_error') + ': ' + str(e))
                self.status.setText(_('error_while_authenticating'))
                self.step = 100
        
        if self.step >= 100:
            # Authenticacion process has concluded
            self.timer.stop()
            self.close()
            return
        elif self.step == 50:
            # Credentials were ok, we check whether we should store them for further uses
            if self.remember:
                self.status.setText(_('storing_credentials'))
                with os.fdopen(os.open(conf.USERCREDSFILE, os.O_WRONLY | os.O_CREAT, 0o600), 'w') as handle:
                    handle.write('[credentials]\nusername=%s\npassword=%s' % (self.uname, encode(self.pw, 'rot_13')))
                    handle.close()
                self.step = 99
            else:
                self.status.setText(_('successfully_authenticated'))
                self.step = 99
            
        self.step = self.step + 1
        self.pbar.setValue(self.step)

    def center(self):
        """
            Description: Just centers the window
            Arguments: None
            Returns: Nothing
        """

        qr = self.frameGeometry()
        cp = QDesktopWidget().availableGeometry().center()
        qr.moveCenter(cp)
        self.move(qr.topLeft())

class Credentials(QDialog):
    """
        This class shows the main credentials Dialog and prompts user for username and
        password. Also a checkbox is shown to store credentials.
    """

    def __init__(self, parent):
        QDialog.__init__(self, parent)
        self.initUI()

    def dismiss(self):
        """
            Description: Dismisses the QDialog box. Called on successful authentication.
            Arguments: None
            Returns: Nothing
        """

        global conf
        if conf.USERNAME:
            self.done(0)

    def check_creds(self):
        """
            Description: Checks whether username or password fields are empty, in which case
                         an error window is shown. Otherwise, try to authenticate vs. oVirt
            Arguments: None
            Returns: Nothing, just dismisses the main window on successful authentication.
        """

        err = QMessageBox()
        uname = self.edit_username.text()
        pw = self.edit_pw.text()

        if not uname or not pw:
            err.critical(self, _('apptitle') + ': ' + _('error'), _('no_empty_creds'))
        else:
            cc = CheckCreds(self, uname, pw, self.remembercreds.isChecked())
            cc.finished.connect(self.dismiss)

    def confirm_quit(self):
        """
            Description: Confirm whether to quit or not the app. This option must be present
                         in case the user cannot authenticate against oVirt, in which case the
                         main window won't be shown but the application will exit instead.
            Arguments: None
            Returns: Nothing, just quits if user confirms.
        """

        cq = QMessageBox.question(self, _('apptitle') + ': ' + _('confirm'), _('confirm_quit'), QMessageBox.Yes | QMessageBox.No, QMessageBox.No)
        if cq == QMessageBox.Yes:
            quit()

    def closeEvent(self, event):
        """
            Description: Deactivation of the red 'x' to close the window. We enforce user to choose
                         whether to authenticate or exit the application.
            Arguments: The event.
            Returns: Nothing
        """

        event.ignore()

    def initUI(self):
        """
            Description: Shows the main dialog, with the username and password fields. Optionally, if
                         allowed by configuration, also shows the "store credentials" checkbox.
            Arguments: None
            Returns: Nothing
        """

        global conf

        self.setFixedSize(400, 150)

        # Keys image
        filename = 'imgs/credentials.png'
        image = QImage(filename)
        imageLabel = QLabel()
        imageLabel.setPixmap(QPixmap.fromImage(image))
        imageLabel.setAlignment(Qt.AlignCenter)

        # Labels for both username and password
        lab_username = QLabel(_('username'))
        lab_username.setAlignment(Qt.AlignRight | Qt.AlignVCenter)
        lab_pw = QLabel(_('password'))
        lab_pw.setAlignment(Qt.AlignRight | Qt.AlignVCenter)

        # LineEdits for both username and password
        self.edit_username = QLineEdit()
        self.edit_username.setFixedWidth(150)
        self.edit_pw = QLineEdit()
        self.edit_pw.setEchoMode(QLineEdit.Password)
        self.edit_pw.setFixedWidth(150)

        # QCheckBox for storing user + password
        self.remembercreds = QCheckBox(_('remember_username_and_password'))

        # OK and cancel buttons
        okButton = QPushButton(_("ok"))
        okButton.setMaximumWidth(100)
        okButton.clicked.connect(self.check_creds)
        cancelButton = QPushButton(_("cancel"))
        cancelButton.setMaximumWidth(100)
        cancelButton.clicked.connect(self.confirm_quit)

        # Grid layout with all the elements
        grid = QGridLayout()

        grid.addWidget(imageLabel, 1, 0, 2, 1)               # Keys image

        grid.addWidget(lab_username, 1, 1, 1, 1)             # Username
        grid.addWidget(self.edit_username, 1, 2, 1, 2)

        grid.addWidget(lab_pw, 2, 1, 1, 1)                   # Password
        grid.addWidget(self.edit_pw, 2, 2, 1, 2)

        if conf.CONFIG['allow_remember'] == '1':
            grid.addWidget(self.remembercreds, 3, 1, 1, 3)   # Remember credentials (only if the option is enabled)

            grid.addWidget(okButton, 4, 1)                   # Buttons
            grid.addWidget(cancelButton, 4, 2)
        else:
            grid.addWidget(okButton, 3, 1)                   # Buttons
            grid.addWidget(cancelButton, 3, 2)
        
        self.setLayout(grid) 

        self.setModal(True)
        self.center()
        self.setWindowIcon(QIcon(IMGDIR + 'credentials.png'))
        self.setWindowTitle(_('apptitle') + ': ' + _('credentials'))
        self.show()

        # If credentials file exists, we'll recover username and password fields
        # and try to authenticate with them
        if isfile(conf.USERCREDSFILE):
            # If credentials file exists and CONFIG['allow_remember'] is disabled, we remove the file
            # as it make no sense keeping it
            if conf.CONFIG['allow_remember'] == '0':
                os.remove(conf.USERCREDSFILE)
            else:
                config = configparser.ConfigParser()
                config.read(conf.USERCREDSFILE)
                uname = config.get('credentials', 'username')
                pw = decode(config.get('credentials', 'password'), 'rot_13')
                self.edit_username.setText(uname)
                self.edit_pw.setText(pw)
                self.check_creds()

    def center(self):
        """
            Description: Just centers the window
            Arguments: None
            Returns: Nothing
        """

        qr = self.frameGeometry()
        cp = QDesktopWidget().availableGeometry().center()
        qr.moveCenter(cp)
        self.move(qr.topLeft())
